### PR TITLE
Treat projections with infer as placeholder during fast reject in new solver

### PR DESCRIPTION
**Pull request by compiler-errors**
r? @lcnr

Kind of a shame that we need to change all of the call sites for `for_each_relevant_impl`, etc. to pass an extra parameter. I guess I could have the "default" fn which calls a configurable fn?

### CODE REVIEW

File: compiler/rustc_middle/src/ty/trait_def.rs [hunk 1]
```diff
@@ -122,12 +122,14 @@ impl<'tcx> TyCtxt<'tcx> {
         self,
         trait_def_id: DefId,
         self_ty: Ty<'tcx>,
+        treat_infer_projections: TreatProjectionsWithInfer,
         mut f: F,
     ) {
```
Comment (lcnr) [inline compiler/rustc_middle/src/ty/trait_def.rs:127]: reminded me to open #108895

File: compiler/rustc_middle/src/ty/fast_reject.rs [hunk 2]
```diff
@@ -53,15 +53,32 @@ pub enum SimplifiedType {
...
+/// inference variables in its substitutions as placeholders or not.
+///
+/// In most cases, unless we know the type is fully normalized or we are handling
+/// the type separately, we should use "NotSimplifyable", as the type cannot be
+/// simplified as if it were rigid.
+#[derive(PartialEq, Eq, Debug, Clone, Copy)]
+pub enum TreatProjectionsWithInfer {
+    AsPlaceholder,
+    NotSimplifyable,
+}
```
Comment (lcnr) [inline compiler/rustc_middle/src/ty/fast_reject.rs:80]: I personally would prefer that this enum talks about the *why*, not the *what*. I think that makes it easier to choose which one to use.

```rust
pub enum TreatProjections {
    /// In the old solver we don't try to normalize projections
    /// when looking up impls and only access them by using the
    /// current self type. This means that if the self type is
    /// a projection which could later be normalized, we must not
    /// treat it as rigid.
    Default,
    /// We can treat projections in the self type as opaque as
    /// we separately look up impls for the normalized self type.
    TraitSolverNextSelfType,
}
```

**Review by lcnr — COMMENTED** [review]
As we share `TreatParams` with the `DeepRejectCtxt`, should we instead add `TreatProjections` separately as another argument to `simplify_type` and keep the current `TreatParams`?

**Comment by lcnr** [comment]
> I guess I could have the "default" fn which calls a configurable fn?

that would work and is probably a bit cleaner :thinking:

**Review by lcnr — COMMENTED** [review]
r=me after nits

**Comment by Mark-Simulacrum** [comment]
This caused a perf regression (see [results](https://github.com/rust-lang/rust/pull/109097#issuecomment-1468047332)). @lcnr / @compiler-errors do we have a sense of whether this was expected or not? Any thoughts on possible fixes?

**Comment by lcnr** [comment]
ahh, that code is really hot so this small change has such a big impact :thinking: 

we should maybe change `simplify_type` to be generic over `TreatParams` and `TreatProjections` instead :thinking:

**Comment by compiler-errors** [comment]
I can take a look at doing that unless lcnr beats me to it

**Comment by nnethercote** [comment]
I am surprised by this regression. I thought the fast reject code was very important for `bitmaps-3.1.0` and not that important for other things. But the numbers don't lie.

**Pull request merged**